lib/configuration.js

```diff
@@ -15,150 +15,615 @@
+ * @memberof Configuration
+ * @public
+ * @function isReady
+ * @returns {Boolean} - returns true if instance initialized without error,
+ *  otherwise false
+ */
+Configuration.prototype.isReady = function() {
+  return this._initState && !this._initError;
+};
+/**
+ * Returns a Boolean indicating whether or not the Configuration instance went
+ * through the initialization process and encountered an error during that
+ * process.
+ * @memberof Configuration
+ * @public
+ * @function hasErrored
+ * @returns {Boolean} - returns true if _initError is set as an Error, otherwise
+ *  false
+ */
+Configuration.prototype.hasErrored = function() {
+  return this._initError instanceof Error;
+};
+/**
+ * The init function causes the Configuration instance to being the
+ * initialization process of gathering configuration from both the metadata
+ * service, runtime configuration and environment. This function will begin the
+ * init process only once. The function will defer starting the initialization
+ * process on any successive calls.
+ * @memberof Configuration
+ * @public
+ * @function init
+ * @chainable
+ * @returns {this} - returns the instance for chaining
+ */
+Configuration.prototype.init = function () {
+  if (this._startedInit === false) {
+    /**
+     * Attempt to get the project number from the metadata service
+     * and kick-off the init process. This would be where we kick off the
+     * `getProjectId` invocation but since that version is not live yet we use
+     * project number for now.
+     * @todo (cristiancavalli) - turn this into the getProjectId call when it
+     *  goes live and remove the getProjectNumber function and call.
+     * @example
+     * utils.getProjectId(this._assimilateProjectId.bind(this));
+     */
+    this._checkLocalServiceContext();
+    utils.getProjectNumber(this._assimilateProjectNumber.bind(this));
+    this._startedInit = true;
   }
+  return this;
+};
+/**
+ * The addReadyListener provides a defined and documented way to attach to the 
+ * `ready` event on a Configuration instance. This interface enforces event
+ * `ready` event registration by using the `once` register handle since the 
+ * `ready` event is only ever supposed to emitted one time. This interface, at
+ * invocation, will also check to see if the `ready` event has already been
+ * emitted at which point it will immediately callback to listener.
+ * @memberof Configuration
+ * @public
+ * @function addReadyListener
+ * @chainable
+ * @param {Configuration~readyCallback} callback - the listener to callback on
+ * @returns {this} - returns the instance for chaining
+ */
+Configuration.prototype.addReadyListener = function(callback) {
+  if (this._initState === false && isFunction(callback)) {
+    // Enforce the one-time fire paradigm
+    this.once('ready', callback);
+  } else if (this._initState === true && isNull(this._initError)) {
+    // Callback immediately as the event already happened
+    callback(this);
```

Comment:
Users may expect that the callback will always fire asynchronously. If you need to callback immediately it is better to invoke inside a `process.nextTick`.